File: java/1_two_sum.java
/** 
    use hash table
    time complexity O(n), space complexity O(n)

    maybe another solution:
        1. sort the nums
        2. two pointers
 */

public static int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> map = new HashMap<>();
    for (int i = 0; i < nums.length; i++) {
        int tmp = target - nums[i];
        if (map.containsKey(tmp)) {
            return new int[]{map.get(tmp), i};
        }
        map.put(nums[i], i);
    }
    throw new IllegalArgumentException("no solution");
}

File: python3/1_Two_Sum.py
'''
Author       : Zhu Honglin
Date         : 2019-05-24 16:44:51
LastEditTime : 2020-09-11 15:10:31
'''

from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nmap = {}
        for idx, val in enumerate(nums):
            tmp = target - val
            if tmp in nmap:
                return [nmap[tmp], idx]
            nmap[val] = idx
        return None

if __name__ == "__main__":
    nums = [3,3]
    target = 6
    a = Solution()
    rst = a.twoSum(nums, target)
    print(rst)